core/src/test/java/org/jobrunr/server/carbonaware/CarbonIntensityForecastTest.java

```diff
@@ -0,0 +1,243 @@
+        for (int i = 0; i < 6; i++) {
+            Instant hour = now.plus(i, HOURS);
+            // Make the current hour the cheapest, increase the rank with each subsequent hour
+            intensityForecast.add(new TimestampedCarbonIntensityForecast(hour, now.plus(i + 1, HOURS), i));
+        }
+        CarbonIntensityForecast carbonIntensityForecast = new CarbonIntensityForecast(new ApiResponseStatus("OK", ""), "ENTSO-E", "10Y1001A1001A82H", "Germany", "Europe/Berlin", Instant.now().plus(1, DAYS), intensityForecast);
+
+        // WHEN
+        Instant result = carbonIntensityForecast.lowestCarbonIntensityInstant(now, to);
+
+        // THEN
+        Instant expected = now.truncatedTo(HOURS);
+        Instant actual = result.truncatedTo(HOURS);
+        assertThat(actual).isEqualTo(expected); // The least expensive hour should be the current hour, truncated to hours for comparison.
+    }
+
+    @Test
+    void lowestCarbonIntensityInstantReturnsNullIfNoHoursBeforeDeadline() {
+        Instant from = Instant.parse("2020-01-01T00:00:00Z");
+        Instant to = Instant.parse("2020-01-01T07:00:00Z");
+        ArrayList<TimestampedCarbonIntensityForecast> intensityForecast = new ArrayList<>(Arrays.asList(
+                new TimestampedCarbonIntensityForecast(Instant.parse("2020-01-01T08:00:00Z"), Instant.parse("2020-01-01T09:00:00Z"), 0),
+                new TimestampedCarbonIntensityForecast(Instant.parse("2020-01-01T09:00:00Z"), Instant.parse("2020-01-01T10:00:00Z"), 1)
+        ));
+        CarbonIntensityForecast carbonIntensityForecast = new CarbonIntensityForecast(new ApiResponseStatus("OK", ""), "ENTSO-E", "10Y1001A1001A82H", "Germany", "Europe/Berlin", Instant.now().plus(1, DAYS), intensityForecast);
+
+        assertThat(carbonIntensityForecast.lowestCarbonIntensityInstant(from, to)).isNull();
+    }
+
+    @Test
+    void lowestCarbonIntensityInstantReturnsNullIfNoHoursAfterFrom() {
+        Instant from = Instant.parse("2020-01-01T08:00:00Z");
+        Instant to = Instant.parse("2020-01-01T10:00:00Z");
+        ArrayList<TimestampedCarbonIntensityForecast> intensityForecast = new ArrayList<>(Arrays.asList(
+                new TimestampedCarbonIntensityForecast(Instant.parse("2020-01-01T06:00:00Z"), Instant.parse("2020-01-01T07:00:00Z"), 0),
+                new TimestampedCarbonIntensityForecast(Instant.parse("2020-01-01T07:00:00Z"), Instant.parse("2020-01-01T08:00:00Z"), 1)
+        ));
+        CarbonIntensityForecast carbonIntensityForecast = new CarbonIntensityForecast(new ApiResponseStatus("OK", ""), "ENTSO-E", "10Y1001A1001A82H", "Germany", "Europe/Berlin", Instant.now().plus(1, DAYS), intensityForecast);
+
+        assertThat(carbonIntensityForecast.lowestCarbonIntensityInstant(from, to)).isNull();
+    }
+
+    @Test
+    void lowestCarbonIntensityInstantWhenFromAndToAreEqual() {
+        Instant from = Instant.parse("2020-01-01T08:00:00Z");
+        Instant to = Instant.parse("2020-01-01T08:00:00Z");
```

Comment:
Is this possible? From == To? I also miss `Returns...` like the other test methods.